### PR TITLE
SIMSBIOHUB 281 - Update Animal Form and Table

### PR DESCRIPTION
**Pull request by GrahamS-Quartech**
## Links to Jira Tickets

- [SIMSBIOHUB-281](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-281)

## Description of Changes

- Removed Critter ID (to be included in edit view later) from Animals table and added WLH ID.
- Added WLH ID to animal submission form.
- Animal ID will now display as Alias
- Separated device list into current devices and previous devices (this is based off deployment timestamps)
- Added collection unit sub section (displays as Ecological Units) to animal submission form. Will correctly disallow duplicates of categories on the same animal.

### CODE REVIEW

File: app/src/constants/i18n.ts [hunk 1]
```diff
@@ -331,11 +331,17 @@ export const SurveyAnimalsI18N = {
...
+  animalCollectionUnitTitle2: 'Ecological Unit',
+  animalCollectionUnitHelp:
+    'Ecological units are groups like population units, herds, and packs. Different species may different units and unit names.',
+  animalCollectionUnitAddBtn: 'Add Unit',
   // Input help strings
   taxonHelp:
     'The species or taxon of the animal. If the species is unknown, select the lowest-ranking known taxon, such as the genus or family',
-  taxonLabelHelp: 'A custom and unique name for you to recognize this individual'
+  taxonLabelHelp: 'A custom and unique name for you to recognize this individual',
+  wlhIdHelp: 'The wildlife health ID associated with this critter.',
```
Comment (curtisupshall) [inline app/src/constants/i18n.ts]: Is "Wildlife Health ID" a proper noun?

File: app/src/constants/i18n.ts [hunk 2]
```diff
@@ -331,11 +331,17 @@ export const SurveyAnimalsI18N = {
...
+  sexHelp: 'The sex of this critter. Leave as default Unknown if unsure.'
```
Comment (curtisupshall) [inline app/src/constants/i18n.ts]: Nit: I would rephrase "Leave as Unknown if unsure."

File: app/src/features/surveys/view/survey-animals/animal.ts [hunk 3]
```diff
@@ -256,6 +269,8 @@ export class Critter {
   critter_id: string;
   taxon_id: string;
   animal_id: string;
+  wlh_id?: string;
+  sex?: string;
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/animal.ts]: Could make this an enum in the future

File: app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  useEffect(() => {
+    if (values.general.taxon_id) {
+      refresh({ route: 'lookups/taxon-collection-categories', query: `taxon_id=${values.general.taxon_id}` });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [values.general.taxon_id]);
+
+  const name: keyof IAnimal = 'collectionUnits';
+  const newCollectionUnit: IAnimalCollectionUnit = {
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx]: I would change this to be a factory function:
```typescript
const createCollectionUnit = (): IAnimalCollectionUnit => ({
    _id: v4(),
    collection_unit_id: '',
    collection_category_id: ''
});
```

File: app/src/components/fields/CbSelectField.tsx [hunk 5]
```diff
@@ -89,7 +90,7 @@ const CbSelectField: React.FC<ICbSelectField> = (props) => {
       {data?.map((a) => {
         const item = typeof a === 'string' ? { label: a, value: a } : { label: a.value, value: a.id };
         return (
-          <MenuItem key={item.value} value={item.value}>
+          <MenuItem disabled={disabledValues?.[item.value]} key={item.value} value={item.value}>
```
Comment (JeremyQuartech) [inline app/src/components/fields/CbSelectField.tsx:93]: Does this still work if the value happens to be 0?

File: app/src/constants/i18n.ts [hunk 6]
```diff
@@ -331,11 +331,17 @@ export const SurveyAnimalsI18N = {
   animalMortalityHelp:
     "Mortality Events describe an individual's death, including the suspected location, date, and cause of death. An individual can only have one Mortality Event.",
   animalMortalityAddBtn: 'Add Mortality',
-
+  animalCollectionUnitTitle: 'Ecological Units',
+  animalCollectionUnitTitle2: 'Ecological Unit',
+  animalCollectionUnitHelp:
+    'Ecological units are groups like population units, herds, and packs. Different species may different units and unit names.',
```
Comment (JeremyQuartech) [inline app/src/constants/i18n.ts]: replace ~'like'~ -> 'such as'

File: app/src/constants/i18n.ts (hunk 1 above)
Comment (JeremyQuartech) [reply app/src/constants/i18n.ts]: Agreed

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx [hunk 7]
```diff
@@ -72,16 +73,30 @@ export const SurveyAnimalsTable = ({
       )
     },
     {
-      field: 'deployments',
-      headerName: 'Device ID',
+      field: 'current_devices',
+      headerName: 'Current Devices',
```
Comment (JeremyQuartech) [inline app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx:77]: Nit: I think 'Active Devices' would sound more natural

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx [hunk 8]
```diff
@@ -72,16 +73,30 @@ export const SurveyAnimalsTable = ({
...
+          (device: IAnimalDeployment) => !device.attachment_end || moment(device.attachment_end).isAfter(moment())
+        );
+        return currentDeploys?.length
+          ? currentDeploys.map((device: IAnimalDeployment) => device.device_id).join(', ')
+          : 'No Devices';
+      }
+    },
+    {
+      field: 'previous_devices',
+      headerName: 'Previous Devices',
```
Comment (JeremyQuartech) [inline app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx:90]: Nit: Maybe previous devices could also be revised. Expired Devices or Inactive Devices?

File: app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    _id: v4(),
+    collection_unit_id: '',
+    collection_category_id: ''
+  };
+
+  //Animals may have multiple collection units, but only one instance of each category.
+  //We use this and pass to the select component to ensure categories already used in the form can't be selected again.
```
Comment (JeremyQuartech) [inline app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx:39]: Is this always true? Could there hypothetically be a transient critter that belongs to multiple 'herds'? I only mention this because I believe critterbase supports that.

File: app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx
Comment (JeremyQuartech) [inline app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx:1]: The behavior is a bit weird if you retroactively change the species after adding a collection unit. Maybe some logic could be added to prevent that.

**Review by JeremyQuartech — COMMENTED** [review]
Nice additions, seems to be working well 👍

File: app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx (hunk 9 above)
Comment (GrahamS-Quartech) [reply app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx:39]: Critterbase would support this but I checked with Macgregor and he wishes to restrict the functionality in this way for right now.

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx (hunk 7 above)
Comment (GrahamS-Quartech) [reply app/src/features/surveys/view/survey-animals/SurveyAnimalsTable.tsx:77]: These terms come from the ticket so I think I will keep them this way unless Macgregor wants something else later.

File: app/src/components/fields/CbSelectField.tsx (hunk 5 above)
Comment (GrahamS-Quartech) [reply app/src/components/fields/CbSelectField.tsx:93]: It should be a string value here, but even if it were the numeric zero that is still a valid index.

File: app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx
Comment (GrahamS-Quartech) [reply app/src/features/surveys/view/survey-animals/form-sections/CollectionUnitAnimalForm.tsx:1]: Good call, I will add something to handle this.

**Review by curtisupshall — APPROVED** [review]

**Review by JeremyQuartech — APPROVED** [review]
